onnxruntime/test/contrib_ops/dynamic_quantize_matmul_test.cc

```diff
@@ -0,0 +1,204 @@
+                                const char* domain = onnxruntime::kMSDomain) : OpTester(op, opset_version, domain),
+                                                                               A_dims_(A_dims),
+                                                                               B_dims_(B_dims),
+                                                                               Y_dims_(Y_dims),
+                                                                               model_file_(model_file) {
+    Init();
+  }
+  void Init() {
+    // create rand inputs
+    RandomValueGenerator random{};
+    A_data_ = random.Uniform<float>(A_dims_, -1.0f, 1.0f);
+
+    std::vector<int> tmp_B_data = random.Uniform<int32_t>(B_dims_, std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+    std::transform(tmp_B_data.begin(), tmp_B_data.end(), std::back_inserter(B_data_), [](int32_t v) -> T {
+      return static_cast<T>(v);
+    });
+
+    B_zero_point_ = {static_cast<T>(random.Uniform<int32_t>({1}, std::numeric_limits<T>::min(), std::numeric_limits<T>::max())[0])};
+    B_scale_ = random.Uniform<float>({1}, -0.1f, 0.1f);
+
+    const int64_t Y_size = std::accumulate(Y_dims_.cbegin(), Y_dims_.cend(), static_cast<int64_t>(1), std::multiplies<int64_t>{});
+    Y_data_.resize(Y_size);
+
+    AddInput<float>("A", A_dims_, A_data_);
+    AddInput<T>("B", B_dims_, B_data_);
+    AddInput<float>("b_scale", {1}, B_scale_);
+    AddInput<T>("b_zero_point", {1}, B_zero_point_);
+
+    AddOutput<float>("Y", Y_dims_, Y_data_);
+  }
+  void Run() {
+#ifndef NDEBUG
+    run_called_ = true;
+#endif
+    std::vector<MLValue> cpu_fetches;
+    std::vector<MLValue> subgraph_fetches;
+    Compute(cpu_fetches);
+    ComputeOriginalSubgraph(subgraph_fetches);
+
+    // Compare CPU with original subgraph on the output(0) - Y
+    ASSERT_TRUE(cpu_fetches.size() >= subgraph_fetches.size());
+    for (size_t i = 0; i < subgraph_fetches.size(); i++) {
+      if (cpu_fetches[i].IsTensor() && subgraph_fetches[i].IsTensor()) {
+        VLOGS_DEFAULT(1) << "Checking tensor " << i;
+        CheckTensor(subgraph_fetches[i].Get<Tensor>(), cpu_fetches[i].Get<Tensor>(), 1e-3, 1e-3);
+      }
+    }
+  }
+
+ private:
+  void Compute(std::vector<MLValue>& cpu_fetches) {
+    auto p_model = BuildGraph();
+    auto& graph = p_model->MainGraph();
+
+    Status status = graph.Resolve();
+    ASSERT_TRUE(status.IsOK()) << status;
+
+    // Hookup the inputs and outputs
+    std::unordered_map<std::string, MLValue> feeds;
+    std::vector<std::string> output_names;
+    FillFeedsAndOutputNames(feeds, output_names);
+
+    SessionOptions so;
+    so.session_logid = op_;
+    so.session_log_verbosity_level = 1;
+
+    RunOptions run_options;
+    run_options.run_tag = op_;
+    run_options.run_log_verbosity_level = 1;
+
+    // run with LayerNormalization
+    InferenceSession layernorm_session_object{so, GetEnvironment()};
+    std::string s1;
```

Comment:
What's the relationship with LayerNormalization? Or is this cut-and-pasted code that needs an update?

Reply:
Did similar thing as in the LayerNormalization. Actually those kind of tests can be generalized. Will setup it in a separate PR.

---
In reply to: [439093681](https://github.com/microsoft/onnxruntime/pull/4195#discussion_r439093681) [](ancestors = 439093681)